tempesta_fw/http_sess.c

```diff
@@ -136,22 +151,60 @@ tfw_http_sticky_redirect_allied(TfwHttpReq *req)
 		&& (req->flags & TFW_HTTP_F_ACCEPT_HTML);
 }
 
+/*
+ * Create redirect mark in following form:
+ *
+ *	<attempt_no> | <timestamp> | HMAC(Secret, attempt_no, timestamp)
+ *
+ * Field <attempt_no> is required to track redirection limit, and <timestamp> is
+ * needed to detect bots, who repeat inital redirect mark in all subsequent
+ * requests.
+ */
+static void
+tfw_http_redir_mark_prepare(RedirMarkVal *mv, char *buf, unsigned int buf_len,
+			    TfwStr *chunks, unsigned int ch_len, TfwStr *rmark)
+{
+	unsigned int att_be32 = cpu_to_be32(mv->att_no);
+	unsigned long ts_be64 = cpu_to_be64(mv->ts);
+	DEFINE_TFW_STR(s_sl, "/");
+	DEFINE_TFW_STR(s_eq, "=");
+
+	bin2hex(buf, &att_be32, sizeof(att_be32));
+	bin2hex(&buf[sizeof(att_be32) * 2], &ts_be64, sizeof(ts_be64));
+	bin2hex(&buf[(sizeof(att_be32) + sizeof(ts_be64)) * 2],
+		mv->hmac, sizeof(mv->hmac));
+
+	bzero_fast(chunks, ch_len);
+	chunks[0] = s_sl;
+	chunks[1] = tfw_cfg_sticky.name;
+	chunks[2] = s_eq;
+	chunks[3].ptr = buf;
+	chunks[3].len = buf_len;
+
+	rmark->ptr = chunks;
+	rmark->len = chunks[0].len;
+	rmark->len += chunks[1].len;
+	rmark->len += chunks[2].len;
+	rmark->len += chunks[3].len;
+	__TFW_STR_CHUNKN_SET(rmark, 4);
+}
+
 static int
-tfw_http_sticky_send_redirect(TfwHttpReq *req, StickyVal *sv)
+tfw_http_sticky_send_redirect(TfwHttpReq *req, StickyVal *sv, RedirMarkVal *mv)
 {
 	unsigned long ts_be64 = cpu_to_be64(sv->ts);
-	TfwStr chunks[3], cookie = { 0 };
+	TfwStr c_chunks[3], m_chunks[4], cookie = { 0 }, rmark = { 0 };
 	DEFINE_TFW_STR(s_eq, "=");
 	TfwHttpResp *resp;
-	char buf[sizeof(*sv) * 2];
+	char c_buf[sizeof(*sv) * 2], m_buf[sizeof(*mv) * 2];
```

Comment:
This function was introduced to expose the size of `RedirMarkVal` structure for `http_parser`. I think that inside of `http_sess.c` it would be more appropriate to depend on the size of the variables than on the structure itself.